services/wallet/controllers_v3.go

```diff
@@ -441,49 +449,144 @@ func LinkUpholdDepositAccountV3(s *Service) func(w http.ResponseWriter, r *http.
+		}
+
+		l := logging.Logger(ctx, "wallet")
+
+		if o := r.Header.Get("Origin"); o != s.dappConf.AllowedOrigin {
+			l.Error().Err(errOriginForbidden).Str("origin", strOr(o, "empty")).Msg("error linking solana address")
+			return handlers.WrapError(errOriginForbidden, "request origin forbidden", http.StatusForbidden)
+		}
+
+		var paymentID inputs.ID
+		if err := inputs.DecodeAndValidateString(ctx, &paymentID, chi.URLParam(r, "paymentID")); err != nil {
+			return handlers.WrapError(err, "invalid paymentID", http.StatusBadRequest)
+		}
+
+		b, err := io.ReadAll(io.LimitReader(r.Body, reqBodyLimit10MB))
+		if err != nil {
+			return handlers.WrapError(err, "error reading body", http.StatusBadRequest)
+		}
+
+		var solReq linkSolanaAddrRequest
+		if err := json.Unmarshal(b, &solReq); err != nil {
+			return handlers.WrapError(err, "error decoding body", http.StatusBadRequest)
+		}
+
+		if _, err := govalidator.ValidateStruct(solReq); err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		if err := s.LinkSolanaAddress(ctx, *paymentID.UUID(), solReq); err != nil {
+			l.Error().Err(err).Msg("error linking solana address")
+
+			var solErr *walletutils.LinkSolanaAddressError
+			switch {
+			case errors.Is(err, model.ErrWalletNotWhitelisted):
+				return handlers.WrapError(model.ErrWalletNotWhitelisted, "rewards wallet not whitelisted", http.StatusForbidden)
+			case errors.Is(err, model.ErrChallengeExpired):
+				return handlers.WrapError(model.ErrChallengeExpired, "linking challenge expired", http.StatusUnauthorized)
+			case errors.Is(err, model.ErrWalletNotFound):
+				return handlers.WrapError(model.ErrWalletNotFound, "rewards wallet not found", http.StatusNotFound)
+			case errors.As(err, &solErr):
+				return handlers.WrapError(solErr, "invalid solana linking message", http.StatusUnauthorized)
+			default:
+				return handlers.WrapError(model.ErrInternalServer, "internal server error", http.StatusInternalServerError)
+			}
+		}
+
+		return handlers.RenderContent(ctx, nil, w, http.StatusOK)
+	}
+}
+
+type challengeRequest struct {
+	PaymentID uuid.UUID `json:"paymentId" valid:"required"`
+}
+
+type challengeResponse struct {
+	Nonce string `json:"challengeId"`
+}
+
+func CreateChallenge(s *Service) handlers.AppHandler {
+	return func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		ctx := r.Context()
+
+		l := logging.Logger(ctx, "wallet")
+
+		b, err := io.ReadAll(io.LimitReader(r.Body, reqBodyLimit10MB))
+		if err != nil {
+			return handlers.WrapError(err, "error reading body", http.StatusBadRequest)
+		}
+
+		var chlReq challengeRequest
+		if err := json.Unmarshal(b, &chlReq); err != nil {
+			return handlers.WrapError(err, "error decoding body", http.StatusBadRequest)
+		}
+
+		if _, err := govalidator.ValidateStruct(chlReq); err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		chl, err := s.CreateChallenge(ctx, chlReq.PaymentID.String())
```

Comment:
The strange thing I've noticed throughout this code base is that, despite us being able to use properly typed UUIDs everywhere (and Postgres supports them via the `uuid-ossp` extension, which we already use), we pass it around as strings, and more often than not store as strings too.

On the database this really impacts performance and storage operations. At runtime, we simply waste computational resources, and potentially use too relaxed types where we can (and should) be stricter.

What if we started making a change, and pass and use properly-typed UUIDs?

Reply:
The `id` used on challenge was not a `uuid` but type `text` this was to accommodate other potential identifiers which are type `text` hence the conversion. It just happens that on occasion the payment id is was being used. This has since been revised and the challenge has been made specific to this use case. see comments on table